server/src/main/java/org/elasticsearch/health/metadata/HealthMetadata.java

```diff
@@ -128,15 +128,22 @@ public record Disk(
         RelativeByteSizeValue highWatermark,
         RelativeByteSizeValue floodStageWatermark,
         RelativeByteSizeValue frozenFloodStageWatermark,
-        ByteSizeValue frozenFloodStageMaxHeadroom
+        ByteSizeValue frozenFloodStageMaxHeadroom,
+
+        ByteSizeValue highMaxHeadroom,
+
+        ByteSizeValue floodStageMaxHeadroom
```

Comment:
I have to admit that I found the initial order here more intuitive. I feel this order is an implementation detail of the serialization/deserialization, so if we could only keep it there we should.

There is an option to do that. You can only add the new fields in the end during serialization and deserialization without changing the order in the constructor.

For example:
```
Disk(
        RelativeByteSizeValue highWatermark,
        ByteSizeValue highMaxHeadroom,
        RelativeByteSizeValue floodStageWatermark,
        ByteSizeValue floodStageMaxHeadroom,
        RelativeByteSizeValue frozenFloodStageWatermark,
        ByteSizeValue frozenFloodStageMaxHeadroom
)
```
And later the parser would do the following:
```
        public static final ConstructingObjectParser<Disk, Void> PARSER = new ConstructingObjectParser<>(
            TYPE,
            true,
            (args) -> new Disk(
                RelativeByteSizeValue.parseRelativeByteSizeValue((String) args[0], HIGH_WATERMARK_FIELD.getPreferredName()),
                ByteSizeValue.parseBytesSizeValue((String) args[4], HIGH_MAX_HEADROOM_FIELD.getPreferredName()),
                RelativeByteSizeValue.parseRelativeByteSizeValue((String) args[1], FLOOD_STAGE_WATERMARK_FIELD.getPreferredName()),
                ByteSizeValue.parseBytesSizeValue((String) args[5], FLOOD_STAGE_MAX_HEADROOM_FIELD.getPreferredName()),
                RelativeByteSizeValue.parseRelativeByteSizeValue((String) args[2], FROZEN_FLOOD_STAGE_WATERMARK_FIELD.getPreferredName()),
                ByteSizeValue.parseBytesSizeValue((String) args[3], FROZEN_FLOOD_STAGE_MAX_HEADROOM_FIELD.getPreferredName())
            )
        );
```
Respectively all the serializers and deserializers would need to be adjusted.

What do you think? This is just a suggestion.

Reply:
Yes I think I can do that @gmarouli , let me try in a next commit.

Reply:
Hmm @gmarouli , in trying it, I have some trouble doing it for the `Disk(StreamInput in) throws IOException` constructor. It does not let me first read from `in` in variables so that I then call the Disk constructor with the correct fields. Any idea how to handle the constructor?

I feel it might be a bit weird to keep some parts of HealthMetadata with the "intuitive" order and some other parts with the needed order for bwc.

Reply:
If you need to do anything more complicated than just calling the inner constructor you'll need to use a factory method `static Disk readFrom(StreamInput in) throws IOException` instead.

Reply:
Ah yes, this doesn't work with the constructor you are absolutely right. We can still create a static method `readFrom` or something like that. I think that's common practice around the code base.

>  feel it might be a bit weird to keep some parts of HealthMetadata with the "intuitive" order and some other parts with the needed order for bwc.

Hm, I see, I never thought of them as a different order because this bwc order would never leak out of the serialization methods and I think that is not that unusual in this situation. WDYT?

Reply:
OK I see! I just added a commit that does it. But I am a bit unsure about how I re-arranged the fields for the XContent from/to parsing. @gmarouli can you check the latest and tell me whether I should not have moved the fields in some places?